docs/proposals/14-wireguard-based-cluster-vpn.md

```diff
@@ -0,0 +1,178 @@
+
+## High Availability
+
+There is an issue if the node that hosts the WireGuard endpoint fails. The endpoint is migrated to another node however the time required to do this might exceed the budget for downtimes although one could argue that a disruption of less than 30 seconds to 1 minute does not qualify as a downtime and will in almost all cases not noticeable by end users. 
+
+In this case we also assume that TCP connections won't be interrupted - they would just appear to hang. We will confirm this as part of the development and testing effort.
+
+As a mitigation we propose to instantiate 2 Kubelink instances in the seed cluster. The instances must run on different nodes (if possible but we assume a proper seed cluster has more than one node). Each shoot cluster connects to both endpoints. This means that the OpenVPN server is reachable with two different IP addresses. The VPN seed sidecars must attempt to connect to both of them and will continue to do so. The "Persistent Keepalive" feature is set to 21 seconds by default but could be reduced. Due to the redundancy this however appears not to be necessary.
+
+It is desirable that both connections are used in an equal manner. One strategy could be to use the kubelink 1 connection if the first target WireGuard address is even (the last byte of the IPv4 address), otherwise the kubelink 2 connection. The `vpn-seed` sidecars can then use the following configuration in their OpenVPN configuration file:
+
+ ```
+<connection>
+remote 192.168.45.3 1194 udp
+</connection>
+ 
+<connection>
+remote 192.168.47.34 1194 udp
+</connection>
+```
+
+OpenVPN will go through the list sequentially and try to connect to these endpoints.
+
+## Load Considerations
+
+What happens in case of a failure.? In this case one router will end up owning all connections as the clients will attempt to use the next connection. This could be mitigated by adding a third redundant WireGuard connection. Using this strategy, the failure of one WireGuard endpoint would result in the equal distribution of connections to the two remaining interfaces. We believe however that this will not be necessary.
+
+The cluster node running the Wireguard endpoint is essentially a router that routes all traffic to the various shoot clusters. This is established and proven technology that already exists since decades and has been highly optimized since then. This is also the technology that hyperscalers rely on to provide VPN connectivity to their customers. This said, hyperscalers essentially provide solutions based on IPsec which is known not to scale as well as Wireguard. Wireguard is a relatively new technology but we have no doubt that it is less stable than existing IPsec solution.
+
+We have some indications but no hard evidence that Wireguard performance and scalability does degrade with more than 1000 peers. This would mean that one single seed cluster should be able to serve roughly 1000 shoot clusters before running into issues. 
```

Comment:
What are those indications?

Reply:
Just a statement from a fellow colleague. No hard evidence.

Reply:
Maybe the colleague can explain and add those indications here (or rather you can add them to the document).

Reply:
I have removed this paragraph and added the reference.